### PR TITLE
Fix hsts support when not using SSL

### LINKED ISSUE #627: HSTS not working using hsts-behind-proxy
Hello,

I'm trying to setup HSTS support using annotations in the ingress but I'm not having luck with it. We have the nginx-ingress in front of an AWS ALB that handles SSL termination.

This is how the ingress looks like:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: nginx
    nginx.org/hsts: "True"
    nginx.org/hsts-behind-proxy: "True"
    nginx.org/hsts-include-subdomains: "True"
    nginx.org/hsts-max-age: "2592000"
    nginx.org/mergeable-ingress-type: master
    nginx.org/redirect-to-https: "True"
  creationTimestamp: 2019-07-03T15:02:57Z
  generation: 1
  name: nginx-host-master-app.dev.company.net
  namespace: default
  resourceVersion: "50956162"
  selfLink: /apis/extensions/v1beta1/namespaces/default/ingresses/nginx-host-master-app.dev.company.net
  uid: a434f4e8-9da3-11e9-9d08-1296d0bc4772
spec:
  rules:
  - host: app.dev.company.net
status:
  loadBalancer:
    ingress:
    - {}
```

I've checked the dynamic configuration that is built by the controller under `/etc/nginx/conf.d/` and the expected directives are not there. 

I verified that changing `nginx.org/proxy-connect-timeout` did actually make a change on the nginx configuration so the annotations are working to a degree.

**Expected behavior**
The HSTS relevant headers are added in the response.

**Your environment**
* Version of the Ingress Controller: 1.5.1-alpine
* Version of Kubernetes: 1.11
* Kubernetes platform: EKS
* Using NGINX or NGINX Plus: NGINX

### PR DESCRIPTION
**Pull request by bilby91**
Closes #627

### Proposed changes
This change will fix the HSTS support when not handling SSL at the ingress layer. Previously, the templating code would only render the HSTS related directives if the `SSL` was true which is never the case when a load balancer that sits in front of the ingress handle the TLS.  

Fixes the following issue: https://github.com/nginxinc/kubernetes-ingress/issues/627

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [X] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have checked that all unit tests pass after adding my changes
- [X] I have updated necessary documentation
- [X] I have rebased my branch onto master
- [X] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

**Comment by pleshakov** [comment]
@bilby91 thanks for the PR! we will review it shortly.

**Comment by bilby91** [comment]
@Dean-Coakley Good catch, I totally missed that. Do you think that moving the block of code on top of `{{if $server.SSL}}` would make sense ? 

BTW, do you have a cheap way of testing this ? The way I tested this took me some time to get stuff setup.

**Comment by Dean-Coakley** [comment]
@bilby91 I think this issue is more complex than we initially expected. I'd recommend waiting until  we can provide you with some guidance on how to meet all edge cases. But feel free to share commits/comments in the meantime if you feel so inclined.


There is no way to test without a cluster - if that's what you're hoping for. However you don't need to test with a cloud loadbalancer. I'm testing against a minikube cluster with the [complete-example](https://github.com/nginxinc/kubernetes-ingress/tree/master/examples/complete-example) deployed. And then use curl `curl -k http://<MINIKUBE>/tea -I -L -H "X-Forwarded-Proto: https"` Perhaps that's easier for you.


You can also run the binary locally against a running cluster. You'd set this up by
1. Compiling the binary `go build -installsuffix cgo -ldflags "-w -X main.version=my-version" -o nginx-ingress github.com/nginxinc/kubernetes-ingress/cmd/nginx-ingress`
2. Run `kubectl proxy` is a separate terminal window
3. Make sure the default server secret is deployed in your cluster
4. Run the binary `./nginx-ingress -proxy 127.0.0.1:8001 -default-server-tls-secret=nginx-ingress/default-server-secret -v=3 -main-template-path=internal/nginx/templates/nginx.tmpl -ingress-template-path=internal/nginx/templates/nginx.ingress.tmpl`

This will help you debug template compilation. However please note _**this will not substitute testing in a cluster**_. There is no nginx running here. You can use this for dev purposes before testing in a cluster.

File: internal/configs/version1/nginx-plus.ingress.tmpl [hunk 1]
```diff
@@ -55,6 +55,14 @@ server {
...
 	{{if not $server.GRPCOnly}}
+	{{- if $server.SSLRedirect}}
+	if ($scheme = http) {
+		return 301 https://$host:{{index $server.SSLPorts 0}}$request_uri;
+	}
+	{{- end}}
+	{{end}}
+	{{- end}}
+
 	{{- if $server.HSTS}}
```
Comment (pleshakov) [inline internal/configs/version1/nginx-plus.ingress.tmpl]: with this suggestion HSTS will only be enabled when TLS termination is enabled and `hsts-behind-proxy` annotation/configmapkey is true.
```suggestion
	{{- if and $server.HSTS (or $server.SSL $server.HSTSBehindProxy)}}
```

Comment (pleshakov) [reply internal/configs/version1/nginx-plus.ingress.tmpl]: Additionally, please move lines 66-78 (the `{{- if and $server.HSTS (or $server.SSL $server.HSTSBehindProxy)}} ... {{end}}` block) above line 56 (`{{if $server.SSL}})`. This will fix the warning message like this `*84 using uninitialized "hsts_header_val" variable`.

Comment (bilby91) [reply internal/configs/version1/nginx-plus.ingress.tmpl]: @pleshakov Thanks for taking the time to review the change! The suggested changes would apply for nginx-plus and regular nginx right ? Just making sure!

Comment (pleshakov) [reply internal/configs/version1/nginx-plus.ingress.tmpl]: yep. for both templates! thanks for double checking.

Comment (bilby91) [reply internal/configs/version1/nginx-plus.ingress.tmpl]: Pushed suggested changes!

**Comment by Dean-Coakley** [comment]
Thanks a lot @bilby91 !
